### ISSUE TITLE
Firebase connection problems from GCP Cloud Run app

### ISSUE THREAD

**Issue by grebenshikov** [firebase/firebase-admin-java#284]
### [REQUIRED] Step 2: Describe your environment

  * Operating System version: GCP Cloud Run (Java 8 runtime). US-Central region.
  * Firebase SDK version: 6.8.1, 6.8.0, 6.7.0
  * Library version: 6.8.1, 6.8.0, 6.7.0
  * Firebase Product: database

### [REQUIRED] Step 3: Describe the problem

I experience intermittent connection problems with Firebase Realtime Database from a Spring Boot App running on Google Cloud Run (Managed) service. It works without any problem locally (tried multiple networks and computers as well), but once deployed as a container on Cloud Run - queries response time to Firebase become unpredictable or sometimes no response at all. 

I can constantly reproduce this problem during last 48+ hours. Not sure if it was the case before. I am under impression that it worked well in the past.

In order to isolate the problem I came up with a simple controller that queries ONE object from a database and waits for the result. Here is the controller code:

@GetMapping("/me")            
        	public String me() throws Exception{	
            	
            	
            	DatabaseReference ref = FirebaseDatabase.getInstance().getReference().child("PromoCodes");
        		logger.info("DatabaseReference: " + ref.getKey());
        		
        		ref.addListenerForSingleValueEvent(new ValueEventListener() {
        			
        			@Override
        			public void onDataChange(DataSnapshot snapshot) {

        				logger.info("-- onDataChange: " + snapshot.getChildrenCount());
        				
        			}
        			
        			@Override
        			public void onCancelled(DatabaseError error) {

        				logger.info("! onCancelled: " + error);
        				
        			}
        		});
        				
        		
        		return "Firebase path: " + ref.getPath();
        	}

Here is one sample response from log files:

2019-07-01 20:22:20.689 PDTGET200 247 B 68 ms Safari 12  /me

2019-07-01 20:27:15.886 PDT2019-07-02 03:27:15.885 WARN 1 --- [database-worker] io.netty.util.internal.MacAddressUtil : Failed to find a usable hardware address from the network interfaces; using random bytes: 8c:e7:67:82:bd:a7:0c:1b

2019-07-01 20:28:28.887 PDT2019-07-02 03:28:28.886 ERROR 1 --- [ebsocket-worker] c.g.f.d.connection.WebsocketConnection : [ws_0] WebSocket error

2019-07-01 20:28:28.887 PDT

2019-07-01 20:28:28.887 PDTjava.lang.IllegalStateException: executor not accepting a task at io.netty.resolver.AddressResolverGroup.getResolver(AddressResolverGroup.java:60) ~[netty-resolver-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.bootstrap.Bootstrap.doResolveAndConnect0(Bootstrap.java:200) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.bootstrap.Bootstrap.access$000(Bootstrap.java:49) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.bootstrap.Bootstrap$1.operationComplete(Bootstrap.java:188) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.bootstrap.Bootstrap$1.operationComplete(Bootstrap.java:174) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:502) [netty-common-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:476) [netty-common-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:415) [netty-common-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.util.concurrent.DefaultPromise.setValue0(DefaultPromise.java:540) [netty-common-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.util.concurrent.DefaultPromise.setSuccess0(DefaultPromise.java:529) [netty-common-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.util.concurrent.DefaultPromise.trySuccess(DefaultPromise.java:101) [netty-common-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.channel.DefaultChannelPromise.trySuccess(DefaultChannelPromise.java:84) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.channel.AbstractChannel$AbstractUnsafe.safeSetSuccess(AbstractChannel.java:1009) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.channel.AbstractChannel$AbstractUnsafe.register0(AbstractChannel.java:517) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.channel.AbstractChannel$AbstractUnsafe.access$200(AbstractChannel.java:428) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.channel.AbstractChannel$AbstractUnsafe$1.run(AbstractChannel.java:487) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163) [netty-common-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:404) [netty-common-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:495) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:905) [netty-common-4.1.34.Final.jar!/:4.1.34.Final] at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) [na:1.8.0_202] at java.util.concurrent.FutureTask.run(Unknown Source) [na:1.8.0_202] at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(Unknown Source) [na:1.8.0_202] at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source) [na:1.8.0_202] at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) [na:1.8.0_202] at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) [na:1.8.0_202] at java.lang.Thread.run(Unknown Source) [na:1.8.0_202]

2019-07-01 20:28:28.887 PDT

2019-07-01 20:29:59.885 PDT2019-07-02 03:29:59.685 INFO 1 --- [se-event-target] ApiApplication : -- onDataChange: 1

In this case I got the response 7+ minutes after the request, including [database-worker] and [ebsocket-worker] exceptions prior to that. This is pretty much a pattern.
Sometimes I get response in a matter of milliseconds like expected, other times it takes 5-15 minutes to get it eventually with a bunch of similar connectivity exceptions in the logs.

#### Steps to reproduce:

Deploy simple Spring Boot app with a controller that queries small object from Firebase to Cloud Run.

Let me know if I can provide any other details to help resolve it.

**Comment by hiranya911** [comment]
If this only happens on Cloud Run, then it's more likely to be a problem with that environment. Have you tried contacting the Cloud Run team with this information?

**Comment by grebenshikov** [comment]
I will contact them for sure. Planning to run the same test on App Engine to isolate Cloud Run infrastructure vs. Firebase location.

**Comment by hiranya911** [comment]
I have multiple test applications on App Engine that uses this library. And we have a few other users who use this API in App Engine for various production apps. Never seen an issue of the reported nature.

**Comment by CodingDoug** [comment]
I had problems using the admin SDK (Firestore) using an image built with openjdk:8-jre-alpine, as cribbed from one of the Google-provided examples for Kotlin/ktor/netty.  After I switched to openjdk:8-jre-stretch, it worked.  The alpine build was a bit too slim and lacked some libraries and classes.  Don't know if that's what you're running into, but I'd give it a try.

**Comment by grebenshikov** [comment]
Thanks for the suggestion about different jdk. I tried it, but it did not solve the problem.

Though I think I have nailed the pattern for this problem:

Step 1. Issue seems to be 100% reproducible then I just start new container (for example re-deploy it).

I get the following three error messages in logs after attempts to read anything from Firebase:

2019-07-02 23:41:57.395 PDT2019-07-03 06:41:57.395 WARN 1 --- [database-worker] io.netty.util.internal.MacAddressUtil : Failed to find a usable hardware address from the network interfaces; using random bytes: 11:7b:e0:13:9f:a5:3e:60

2019-07-02 23:43:33.196 PDT2019-07-03 06:43:33.195 ERROR 1 --- [ebsocket-worker] c.g.f.d.connection.WebsocketConnection : [ws_0] WebSocket error

2019-07-02 23:43:33.196 PDT

2019-07-02 23:43:33.196 PDTjava.lang.IllegalStateException: executor not accepting a task at io.netty.resolver.AddressResolverGroup.getResolver(AddressResolverGroup.java:60) ~[netty-resolver-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.bootstrap.Bootstrap.doResolveAndConnect0(Bootstrap.java:200) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.bootstrap.Bootstrap.access$000(Bootstrap.java:49) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.bootstrap.Bootstrap$1.operationComplete(Bootstrap.java:188) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.bootstrap.Bootstrap$1.operationComplete(Bootstrap.java:174) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:502) [netty-common-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:476) [netty-common-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:415) [netty-common-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.util.concurrent.DefaultPromise.setValue0(DefaultPromise.java:540) [netty-common-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.util.concurrent.DefaultPromise.setSuccess0(DefaultPromise.java:529) [netty-common-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.util.concurrent.DefaultPromise.trySuccess(DefaultPromise.java:101) [netty-common-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.channel.DefaultChannelPromise.trySuccess(DefaultChannelPromise.java:84) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.channel.AbstractChannel$AbstractUnsafe.safeSetSuccess(AbstractChannel.java:1009) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.channel.AbstractChannel$AbstractUnsafe.register0(AbstractChannel.java:517) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.channel.AbstractChannel$AbstractUnsafe.access$200(AbstractChannel.java:428) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.channel.AbstractChannel$AbstractUnsafe$1.run(AbstractChannel.java:487) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163) [netty-common-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:404) [netty-common-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:495) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:905) [netty-common-4.1.34.Final.jar!/:4.1.34.Final] at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [na:1.8.0_212] at java.util.concurrent.FutureTask.run(FutureTask.java:266) [na:1.8.0_212] at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) [na:1.8.0_212] at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) [na:1.8.0_212] at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [na:1.8.0_212] at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [na:1.8.0_212] at java.lang.Thread.run(Thread.java:748) [na:1.8.0_212]


Step 2. It take 5-10 minutes for query result to eventually return results (onDataChange callbacks). It happens after the exceptions above:

2019-07-02 23:45:28.595 PDT2019-07-03 06:45:28.595 INFO 1 --- [se-event-target] ApiApplication : -- onDataChange: 1

Step 3. Right after that moment system works perfect. Expected quick responses from Firebase for new requests... 


Step 4. Wait for container to be restarted due to inactivity (or just re-deploy app one more time). And the story repeats from #1 above.

Looks like it's the stable pattern to reproduce that problem. 
BTW - it works like a charm on App Engine.

P.S. I will try to report this issue to Cloud Run, but I may be you guys can try to escalate it from your (Firebase) side as well.


Thanks!

**Comment by hiranya911** [comment]
Thanks for the detailed analysis @grebenshikov. We will try to follow up with the Cloud Run team via internal channels as well.

Can you share your test case (source code) with us? Ideally a simple web app and a Docker file, or even a hosted container that we can inspect and test?

**Comment by grebenshikov** [comment]
@hiranya911 - please contact me directly and I would be happy to provide you details or even access to my hosted container to help you debug the problem. So far I can confirm that pattern continue to remain THE SAME over the week. The first Firebase query causes exceptions and several minutes delay before it "self-heals" and works fine till next container restart.

What is interesting - I saw one more detailed exception today that refers to you guys.
**Uncaught exception in Firebase Database runloop (6.8.1). Please report to firebase-database-client@google.com**

Below is the full stack trace with that exception.

P.S. I have no heard back from Cloud Run team on my ticket...


2019-07-09 21:06:05.262 PDT2019-07-10 04:06:04.863 WARN 1 --- [database-worker] io.netty.util.internal.MacAddressUtil : Failed to find a usable hardware address from the network interfaces; using random bytes: 9a:bd:99:d5:71:db:e4:13

2019-07-09 21:07:26.663 PDT2019-07-10 04:07:26.465 ERROR 1 --- [ebsocket-worker] c.g.f.d.connection.WebsocketConnection : [ws_0] WebSocket error

2019-07-09 21:07:26.663 PDT


2019-07-09 21:07:26.663 PDTjava.lang.IllegalStateException: executor not accepting a task at io.netty.resolver.AddressResolverGroup.getResolver(AddressResolverGroup.java:60) ~[netty-resolver-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.bootstrap.Bootstrap.doResolveAndConnect0(Bootstrap.java:200) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.bootstrap.Bootstrap.access$000(Bootstrap.java:49) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.bootstrap.Bootstrap$1.operationComplete(Bootstrap.java:188) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.bootstrap.Bootstrap$1.operationComplete(Bootstrap.java:174) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:502) [netty-common-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:476) [netty-common-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:415) [netty-common-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.util.concurrent.DefaultPromise.setValue0(DefaultPromise.java:540) [netty-common-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.util.concurrent.DefaultPromise.setSuccess0(DefaultPromise.java:529) [netty-common-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.util.concurrent.DefaultPromise.trySuccess(DefaultPromise.java:101) [netty-common-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.channel.DefaultChannelPromise.trySuccess(DefaultChannelPromise.java:84) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.channel.AbstractChannel$AbstractUnsafe.safeSetSuccess(AbstractChannel.java:1009) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.channel.AbstractChannel$AbstractUnsafe.register0(AbstractChannel.java:517) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.channel.AbstractChannel$AbstractUnsafe.access$200(AbstractChannel.java:428) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.channel.AbstractChannel$AbstractUnsafe$1.run(AbstractChannel.java:487) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163) [netty-common-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:404) [netty-common-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:495) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final] at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:905) [netty-common-4.1.34.Final.jar!/:4.1.34.Final] at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [na:1.8.0_212] at java.util.concurrent.FutureTask.run(FutureTask.java:266) [na:1.8.0_212] at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) [na:1.8.0_212] at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) [na:1.8.0_212] at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [na:1.8.0_212] at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [na:1.8.0_212] at java.lang.Thread.run(Thread.java:748) [na:1.8.0_212]

2019-07-09 21:07:26.663 PDT


2019-07-09 21:09:16.763 PDT2019-07-10 04:09:16.762 ERROR 1 --- [database-worker] c.google.firebase.database.core.RunLoop : Uncaught exception in Firebase Database runloop (6.8.1). Please report to firebase-database-client@google.com

2019-07-09 21:09:16.763 PDT


2019-07-09 21:09:16.763 PDTjava.lang.NullPointerException: null at com.google.firebase.database.connection.WebsocketConnection$WSClientHandlerImpl$1.run(WebsocketConnection.java:273) ~[firebase-admin-6.8.1.jar!/:na] at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[na:1.8.0_212] at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[na:1.8.0_212] at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) ~[na:1.8.0_212] at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) ~[na:1.8.0_212] at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[na:1.8.0_212] at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [na:1.8.0_212] at java.lang.Thread.run(Thread.java:748) [na:1.8.0_212]

**Comment by hiranya911** [comment]
Can you share the source code please? Just zip up a sample webapp project that we can build and deploy.

**Comment by grebenshikov** [comment]
Sure. Here it is - https://www.dropbox.com/s/o6o55p3xeyy8gih/cloudrun-firebase-test.zip?dl=0

Look at "readme.txt" for instructions to configure it for your environment.

I tested it as newly deployed container @Cloud Run. Issue is 100% reproducible.

**Comment by hiranya911** [comment]
@grebenshikov I can't reproduce the issue. Just uploaded your project to Cloud Run and made some calls with curl. The longest cold start time I observed was around 18s which is to be expected. But nothing as outrageous as 7 minutes. 

```
2019-07-15 16:42:07.434 PDT /\\ / ___'_ __ _ _(_)_ __ __ _ \ \ \ \
2019-07-15 16:42:07.434 PDT( ( )\___ | '_ | '_| | '_ \/ _` | \ \ \ \
2019-07-15 16:42:07.434 PDT \\/ ___)| |_)| | | | | || (_| | ) ) ) )
2019-07-15 16:42:07.434 PDT ' |____| .__|_| |_|_| |_\__, | / / / /
2019-07-15 16:42:07.434 PDT =========|_|==============|___/=/_/_/_/
2019-07-15 16:42:07.438 PDT :: Spring Boot :: (v2.1.4.RELEASE)
2019-07-15 16:42:07.438 PDT
2019-07-15 16:42:08.132 PDT2019-07-15 23:42:08.121 INFO 1 --- [ main] c.myapp.CloudrunFirebaseTestApplication : Starting CloudrunFirebaseTestApplication v0.0.1-SNAPSHOT on localhost with PID 1 (/cloudrun-firebase-test.jar started by root in /)
2019-07-15 16:42:08.135 PDT2019-07-15 23:42:08.134 INFO 1 --- [ main] c.myapp.CloudrunFirebaseTestApplication : No active profile set, falling back to default profiles: default
2019-07-15 16:42:15.232 PDT2019-07-15 23:42:15.231 INFO 1 --- [ main] o.s.b.w.embedded.tomcat.TomcatWebServer : Tomcat initialized with port(s): 8080 (http)
2019-07-15 16:42:15.405 PDT2019-07-15 23:42:15.405 INFO 1 --- [ main] o.apache.catalina.core.StandardService : Starting service [Tomcat]
2019-07-15 16:42:15.405 PDT2019-07-15 23:42:15.405 INFO 1 --- [ main] org.apache.catalina.core.StandardEngine : Starting Servlet engine: [Apache Tomcat/9.0.17]
2019-07-15 16:42:15.954 PDT2019-07-15 23:42:15.954 INFO 1 --- [ main] o.a.c.c.C.[Tomcat].[localhost].[/] : Initializing Spring embedded WebApplicationContext
2019-07-15 16:42:15.955 PDT2019-07-15 23:42:15.954 INFO 1 --- [ main] o.s.web.context.ContextLoader : Root WebApplicationContext: initialization completed in 7428 ms
2019-07-15 16:42:16.593 PDT2019-07-15 23:42:16.593 INFO 1 --- [ main] com.myapp.FirebaseBean : --- FirebaseInitBean has been initialized successfully | https://fireflicks-io.firebaseio.com
2019-07-15 16:42:17.313 PDT2019-07-15 23:42:17.313 INFO 1 --- [ main] o.s.s.concurrent.ThreadPoolTaskExecutor : Initializing ExecutorService 'applicationTaskExecutor'
2019-07-15 16:42:18.596 PDT2019-07-15 23:42:18.595 INFO 1 --- [ main] o.s.b.w.embedded.tomcat.TomcatWebServer : Tomcat started on port(s): 8080 (http) with context path ''
2019-07-15 16:42:18.606 PDT2019-07-15 23:42:18.605 INFO 1 --- [ main] c.myapp.CloudrunFirebaseTestApplication : Started CloudrunFirebaseTestApplication in 12.998 seconds (JVM running for 16.296)
2019-07-15 16:42:18.929 PDT2019-07-15 23:42:18.928 INFO 1 --- [nio-8080-exec-1] o.a.c.c.C.[Tomcat].[localhost].[/] : Initializing Spring DispatcherServlet 'dispatcherServlet'
2019-07-15 16:42:18.930 PDT2019-07-15 23:42:18.929 INFO 1 --- [nio-8080-exec-1] o.s.web.servlet.DispatcherServlet : Initializing Servlet 'dispatcherServlet'
2019-07-15 16:42:18.996 PDT2019-07-15 23:42:18.996 INFO 1 --- [nio-8080-exec-1] o.s.web.servlet.DispatcherServlet : Completed initialization in 67 ms
2019-07-15 16:42:19.323 PDT2019-07-15 23:42:19.323 INFO 1 --- [nio-8080-exec-1] com.myapp.MyController : DatabaseReference: PromoCodes
2019-07-15 16:42:19.526 PDTGET200 259 B 18.3 s curl/7.53.1  /me
2019-07-15 16:42:26.138 PDT2019-07-15 23:42:26.138 INFO 1 --- [nio-8080-exec-4] com.myapp.MyController : DatabaseReference: PromoCodes
2019-07-15 16:42:26.142 PDTGET200 251 B 22 ms curl/7.53.1  /me
2019-07-15 16:45:42.009 PDT2019-07-15 23:45:42.008 INFO 1 --- [nio-8080-exec-8] com.myapp.MyController : DatabaseReference: PromoCodes
2019-07-15 16:45:42.012 PDTGET200 250 B 11 ms curl/7.53.1  /me
```

From the above log you can see that the first request took 18.3 seconbds (around 12s was that spent on the Sprint/Tomcat/framework initialization).

**Comment by hiranya911** [comment]
FWIW, I do see the error messages you mentioned, but they seem to occur after a period of inactivity -- likely due to Cloud Run terminating the running container instance. For example, roughly 3 minutes after my last request I see:

```
2019-07-15 16:48:33.184 PDT2019-07-15 23:48:33.183 WARN 1 --- [database-worker] io.netty.util.internal.MacAddressUtil : Failed to find a usable hardware address from the network interfaces; using random bytes: 0d:61:cb:4e:a0:e1:09:28
2019-07-15 16:50:02.484 PDT2019-07-15 23:50:01.884 ERROR 1 --- [ebsocket-worker] c.g.f.d.connection.WebsocketConnection : [ws_0] WebSocket error
```

**Comment by hiranya911** [comment]
I have a feeling your notion of "response time" is based on when the RTDB event listener logs the data read from the database:

```
2019-07-15 16:51:45.084 PDT2019-07-15 23:51:45.084 INFO 1 --- [se-event-target] com.myapp.MyController : -- onDataChange: getChildrenCount=0
```

Is that right? In that case please note that RTDB event listeners are asynchronous and run on a separate thread. How Cloud Run handles the execution of such async threads (specifically, how their logs are collected and presented in the console) are completely undeterministic. You cannot rely on that timestamp to reason about the response time. For that you need to halt your web request thread, until the event listener fires.

**Comment by grebenshikov** [comment]
Hey, thanks for testing my test app and sharing your observations. I think you are directionally correct. The way Cloud Run handles async threads (AND not only the way logs are collected - that one I am not sure if this adds up to the problem), but it for sure "plays" some games with infrastructure resources if no web workers are active... In the end it's designed to handle HTTP requests and control container lifetime during its execution.

I did a simple change of the code as you suggested - added HALT to my web request thread while event listener does it job... and problem seems to go away!!! Now I got completely synchronous way to work with RTDB from Cloud Run and it looks effective as well.
For example for the first execution of my request, including time to query RTDB it took 2.3sec (after container start up time) and following requests are processed in 15-20ms timeframe.
